README.rst
chip_python_aREST
============================
A REST API for controlling CHIP GPIO

Manual::

For Python2.7::

    sudo apt-get update
    sudo apt-get install git build-essential python-dev python-pip flex bison python-flask -y
    git clone https://github.com/xtacocorex/chip_python_aREST.git
    cd chip_python_aREST
    sudo python setup.py install

For Python3::

    sudo apt-get update
    sudo apt-get install git build-essential python3-dev python3-pip flex bison python-flask -y
    git clone https://github.com/xtacocorex/chip_python_aREST.git
    cd chip_python_aREST
    sudo python3 setup.py install

CHIP_IO is required to use this library, code and instructions are here: https://github.com/xtacocorex/CHIP_IO
Scripts using this library will need to be run with root permissions (sudo or started at boot by init script).

This library is also available on pypi and can be installed via pip/pip3.:

    # Python2
    sudo pip install chip-python-aREST
    # Python3
    sudo pip3 install chip-python-aREST

Usage
--------

**Example Script**

The following is an example script that details a basic way to instantiate the REST API on your CHIP::

    import CHIP_aREST.aREST as aREST

    # Setup the id
    # The id is a special identifier for your CHIP
    aREST.set_id('5gad42')

    # Setup the name
    # The name can be anything you want
    aREST.set_name("My Local CHIP")

    # Setup the hardware type
    # Not really needed until CHIP_IO get CHIP Pro support
    aREST.set_hardware("chip")

    # This is where any variables and functions would be setup

    # Start the API
    # Debug can be turned on/off
    # Keep the host 0.0.0.0 to allow for local network access
    # Port can be whatever you want it to be
    aREST.RestApp(host="0.0.0.0"<Plug>PeepOpenort=3000,debug=True)

The API also supports user specified variables and functions::

    # First create the variable
    temperature = 25.2
    # Then add it to the API
    aREST.variable("temperature",temperature)

    # For funtions, we need to define it first
    # Functions can have arguments and they can be fed in with url parameters
    def myfunction():
        # you can do whatever you want here
        # CHIP_IO Specific calls
        # crazy math
        # or in this case
        return "myfunction was called, howdy!"

    # Now we add it to the API
    # Make sure you don't add the () to the function
    aREST.function("functiontest",myfunction)

Cloud features TBD.

Example scripts are found in the examples folder.  They are also installed into /usr/local/bin/.

REST API
---------

For local instances of the API, you can access the CHIP via:

    http://192.168.0.5:3000/

Replace the IP address with the one for your CHIP.  If you have avahi installed on your CHIP, you can replace the IP address with <hostname>.local.

All of the REST API are detailed in the tables below.  Note the HTTP Method used for the call.  Not everything uses a normal HTTP GET method.

If you use a web browser to send the URL to the CHIP, you are limited to the GET method.

The curl program installed in Linux or MacOS/OS X can be used to test the API::

    curl -X  GET http://chipdev.local:3000/digital/csid0/1
    curl -X  GET http://chipdev.local:3000/temperature
    curl -X  PUT http://chipdev.local:3000/temperature?value=-24.2

**Basics**

  +------------------+-------------------------------+------------------------------------------------------+
  | **Method**       | **Command**                   | **Description**                                      |
  +------------------+-------------------------------+------------------------------------------------------+
  | GET              | /                             | Basic Info                                           |
  +------------------+-------------------------------+------------------------------------------------------+
  | GET              | /id                           | Basic Info                                           |
  +------------------+-------------------------------+------------------------------------------------------+
  | GET              | /debug                        | Dump all data                                        |
  +------------------+-------------------------------+------------------------------------------------------+
  | GET              | /<variablename>               | Get value of <variablename>                          |
  +------------------+-------------------------------+------------------------------------------------------+
  | DELETE           | /<variablename>               | Delete <variablename>                                |
  +------------------+-------------------------------+------------------------------------------------------+
  | PUT, POST        | /<variablename>?value=<value> | Set <variablename> to <value>                        |
  +------------------+-------------------------------+------------------------------------------------------+
  | GET              | /<functionname>               | Get value of <functionname>                          |
  +------------------+-------------------------------+------------------------------------------------------+
  | GET              | /<functionname>?value=<value> | Get value of <functionname> with <value> as input    |
  +------------------+-------------------------------+------------------------------------------------------+
  | GET              | /version                      | Get current CHIP_IO version                          |
  +------------------+-------------------------------+------------------------------------------------------+

**GPIO**

  +------------------+-------------------------------+------------------------------------------------------+
  | **Method**       | **Command**                   | **Description**                                      |
  +------------------+-------------------------------+------------------------------------------------------+
  | GET, PUT, POST   | /digital/<pinname>/[0,1]      | Digital Write 0 or 1 to <pinname>                    |
  +------------------+-------------------------------+------------------------------------------------------+
  | GET              | /digital/<pinname>            | Digital Read <pinname>                               |
  +------------------+-------------------------------+------------------------------------------------------+
  | GET              | /digital/<pinname>/r          | Digital Read <pinname>                               |
  +------------------+-------------------------------+------------------------------------------------------+
  | GET              | /digital/cleanup              | Cleanup all GPIO Pins                                |
  +------------------+-------------------------------+------------------------------------------------------+
  | GET              | /digital/cleanup/<pinname>    | Cleanup only GPIO Pin: <pinname>                     |
  +------------------+-------------------------------+------------------------------------------------------+

**PWM**

  +------------------+-----------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------+
  | **Method**       | **Command**                                               | **Description**                                                                                                   |
  +------------------+-----------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------+
  | GET              | /pwm/0/start?duty_cycle=[0.0-100.0]&frequency=<frequency> | Start PWM0 with duty cycle and frequency.  duty_cycle and frequency are optional, they default to 25.0% and 200.0 |
  +------------------+-----------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------+
  | GET              | /pwm/0/stop                                               | Stop PWM0                                                                                                         |
  +------------------+-----------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------+
  | GET              | /pwm/0/cleanup                                            | Cleanup PWM0                                                                                                      |
  +------------------+-----------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------+
  | GET, PUT, POST   | /pwm/0/duty_cycle/[0.0-100.0]                             | Change PWM0 Duty Cycle                                                                                            |
  +------------------+-----------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------+
  | GET, PUT, POST   | /pwm/0/frequency/<frequency>                              | Change PWM0 Frequency                                                                                             |
  +------------------+-----------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------+

**Software PWM**

  +------------------+-----------------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------+
  | **Method**       | **Command**                                                           | **Description**                                                                                                                   |
  +------------------+-----------------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------+
  | GET              | /softpwm/<pinname>/start?duty_cycle=[0.0-100.0]&frequency=<frequency> | Start SoftPWM on <pinname> with duty cycle and frequency.  duty_cycle and frequency are optional, they default to 25.0% and 35.0  |
  +------------------+-----------------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------+
  | GET              | /softpwm/<pinname>/stop                                               | Stop SoftPWM on <pinname>                                                                                                         |
  +------------------+-----------------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------+
  | GET              | /softpwm/<pinname>/cleanup                                            | Cleanup SoftPWM on <pinname>                                                                                                      |
  +------------------+-----------------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------+
  | GET, PUT, POST   | /softpwm/<pinname>/duty_cycle/[0.0-100.0]                             | Change SoftPWM Duty Cycle on <pinname>                                                                                            |
  +------------------+-----------------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------+
  | GET, PUT, POST   | /softpwm/<pinname>/frequency/<frequency>                              | Change SoftPWM Frequency on <pinname>                                                                                             |
  +------------------+-----------------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------+

**LRADC**

  +------------------+-------------------------------------------+------------------------------------------------------+
  | **Method**       | **Command**                               | **Description**                                      |
  +------------------+-------------------------------------------+------------------------------------------------------+
  | GET, PUT, POST   | /analog/sample_rate/[32.25,62.5,125,250]  | Set LRADC Sample Rate to 32.25, 62.5, 125, or 250    |
  +------------------+-------------------------------------------+------------------------------------------------------+
  | GET              | /analog/sample_rate                       | Get currrent LRADC Sample Rate                       |
  +------------------+-------------------------------------------+------------------------------------------------------+
  | GET              | /analog/scale_factor                      | Get LRADC Scale Factor                               |
  +------------------+-------------------------------------------+------------------------------------------------------+
  | GET              | /analog/raw/[0,1]                         | Get raw LRADC output for channel 0 or 1              |
  +------------------+-------------------------------------------+------------------------------------------------------+
  | GET              | /analog/full/[0,1]                        | Get full LRADC output for channel 0 or 1             |
  +------------------+-------------------------------------------+------------------------------------------------------+

  **Utilities**

  +------------------+-------------------------------------------+------------------------------------------------------+
  | **Method**       | **Command**                               | **Description**                                      |
  +------------------+-------------------------------------------+------------------------------------------------------+
  | GET, PUT, POST   | /1v8_pin/enable/[1.8,2.0,2.6,3.3]         | Enable 1.8V Pin to output 1.8, 2.0, 2.6, or 3.3V     |
  +------------------+-------------------------------------------+------------------------------------------------------+
  | GET              | /1v8_pin/voltage                          | Get currrent 1.8V Pin voltage setting                |
  +------------------+-------------------------------------------+------------------------------------------------------+
  | GET              | /1v8_pin/disable                          | Disable 1.8V Pin                                     |
  +------------------+-------------------------------------------+------------------------------------------------------+
  | GET              | /unexport_all                             | Backup function to unexport all GPIO                 |
  +------------------+-------------------------------------------+------------------------------------------------------+

Credits
--------

Marco Schwartz for the original pi-aREST node.js code that was used as a basis for this code.
(https://github.com/marcoschwartz/pi-aREST)

License
-------

CHIP Python aREST by Robert Wolterman, released under the MIT License.


CHANGELOG.rst
0.3.0
----
* SERVO support
* CHIP Pro support

0.2.1
----
* Python3 Fixes

0.2.0
----
* Finished the encapsulation
* PWM and SoftPWM added
* Variable Delete
* Functions verified
* README updates

0.1.3
----
* Encapsulated everything into the class since it keeps everything cleaner
  - By doing this, I cleaned up the actual Flask decorator code

0.1.2
----
* Added per pin cleanup code
* Added debug hook to return all the data

0.1.1
----
* Added digital read/write capability
* Added digital cleanup capability

0.1.0
----
* Initial stab at getting this to work
  - REST API info
  - Custom Variables
  - Custom Functions (untested)
  - CHIP_IO Version
  - CHIP_IO LRADC control and status
  - CHIP_IO Utilities for 1.8VDC Pin Control and unexporting all the GPIO Pins
* Basic example script (gets installed to /usr/bin)


CHIP_aREST/aREST.py
#!flask/bin/python

# aREST style API for controlling the CHIP via JSON and CHIP_IO
# Robert Wolterman, 2016
#
# Port from node.js to Python
# https://github.com/marcoschwartz/pi-aREST/blob/master/index.js
#
# Analog code adapted from:
# https://github.com/marcoschwartz/aREST-ESP8266/blob/master/aREST.lua
#
# All other code I came up with to mimic the style

# Module Imports
from flask import Flask, jsonify, request
import CHIP_IO.GPIO as GPIO
import CHIP_IO.PWM as PWM
import CHIP_IO.SOFTPWM as SPWM
import CHIP_IO.SERVO as SERVO
import CHIP_IO.LRADC as LRADC
import CHIP_IO.OverlayManager as OM
import CHIP_IO.Utilities as UT
import random
import math
import copy
import sys
import signal
import paho.mqtt.client as mqtt

# Signal Handling
def sig_handler(signal, frame):
    if not UT.is_chip_pro():
        OM.unload("PWM0")
        UT.disable_1v8_pin()
    GPIO.cleanup()
    PWM.cleanup()
    SPWM.cleanup()
    SERVO.cleanup()
    sys.exit(0)

signal.signal(signal.SIGTERM, sig_handler)

# Classes
class CHIP_RestAPI(Flask):
    def __init__(self, *args, **kwargs):
        super(CHIP_RestAPI, self).__init__(*args, **kwargs)
        # Figure out if we're a CHIP or CHIP Pro automatically
        hw = "chip"
        if UT.is_chip_pro():
            hw = "chip pro"
        # Build initial CHIP_INFO
        self.CHIP_INFO = {
            "id" : "001",
            "name" : "CHIPDEV",
            "hardware" : hw,
            "connected" : False
        }
        self.VARIABLES = {}
        self.FUNCTIONS = {}
        self.PINS_IN_USE = []

    def make_id(self,mylen):
        text = ""
        possible = "abcdefghijklmnopqrstuvwxyz0123456789"
        for i in xrange(mylen):
            text += possible[int(math.floor(random.random() * len(possible)))]
        return text

    def set_id(self,id):
        self.CHIP_INFO["id"] = id

    def get_id(self):
        return self.CHIP_INFO["id"]

    def set_name(self,name):
        self.CHIP_INFO["name"] = name

    def get_name(self):
        return self.CHIP_INFO["name"]

    def set_hardware(self,hw):
        self.CHIP_INFO["hardware"] = hw

    def get_hardware(self):
        return self.CHIP_INFO["hardware"]

    def set_variable(self,name,value):
        self.VARIABLES[name] = value

    def get_variable(self,name):
        if name in self.VARIABLES:
            return self.VARIABLES[name]
        else:
            return -1

    def set_function(self,name,funct):
        self.FUNCTIONS[name] = funct

    def api_debug(self):
        resp = copy.deepcopy(self.CHIP_INFO)
        resp["connected"] = True
        resp["message"] = "Debug Data"
        resp["variables"] = self.VARIABLES
        resp["functions"] = self.FUNCTIONS
        resp["pins in use"] = self.PINS_IN_USE
        return jsonify(resp)

    def api_chipio_version(self):
        resp = copy.deepcopy(self.CHIP_INFO)
        resp["connected"] = True
        resp["message"] = GPIO.VERSION
        return jsonify(resp)

    def api_get_variables(self,variable,req_method,req_args):
        resp = copy.deepcopy(self.CHIP_INFO)
        resp["connected"] = True

        if variable in self.VARIABLES:
            if req_method == 'GET':
                resp["connected"] = True
                resp[variable] = self.VARIABLES[variable]
            elif req_method in ['PUT','POST']:
                value = req_args.get('value')
                resp["connected"] = True
                self.VARIABLES[variable] = value
                resp[variable] = value
            elif req_method == 'DELETE':
                tmp = self.VARIABLES.pop(variable,None)
                resp["message"] = "Variable {0} deleted".format(variable)
                
        if variable in self.FUNCTIONS:
            #  Handle function arguments
            if req_args:
                ddict = req_args.to_dict()
                rtn = self.FUNCTIONS[variable](**ddict)
                resp["return_value"] = rtn
            else:
                rtn = self.FUNCTIONS[variable]()
                resp["return_value"] = rtn

        return jsonify(resp)

    def api_digital_cleanup(self):
        resp = copy.deepcopy(self.CHIP_INFO)
        resp["connected"] = True
        resp["message"] = "All GPIO pins cleaned up"

        self.PINS_IN_USE = []
        GPIO.cleanup()

        return jsonify(resp)

    def api_digital_pin_cleanup(self,pin):
        resp = copy.deepcopy(self.CHIP_INFO)
        resp["connected"] = True

        if not self.PINS_IN_USE:
            resp["message"] = "No pins currently setup"
        else:
            pin = pin.upper()

            if pin not in self.PINS_IN_USE:
                resp["message"] = "Pin not previously in use"
            else:
                resp["message"] = "Cleaning up %s" % pin
                GPIO.cleanup(pin)
                self.PINS_IN_USE.remove(pin)

        return jsonify(resp)

    def api_digital_write(self,pin,value):
        resp = copy.deepcopy(self.CHIP_INFO)
        resp["connected"] = True

        pin = pin.upper()

        # Setup pin if it isn't already and then add it
        if pin not in self.PINS_IN_USE:
            GPIO.setup(pin,GPIO.OUT)
            self.PINS_IN_USE.append(pin)

        # Write data to the pin
        if value == 0:
            resp["message"] = "Writing 0 to " + pin
            GPIO.output(pin,GPIO.LOW)
        elif value == 1:
            resp["message"] = "Writing 1 to " + pin
            GPIO.output(pin,GPIO.HIGH)
        else:
            resp["message"] = "Invalid value specified for " + pin

        return jsonify(resp)

    def api_digital_read(self,pin):
        resp = copy.deepcopy(self.CHIP_INFO)
        resp["connected"] = True

        pin = pin.upper()

        # Setup pin if it isn't already and then add it
        if pin not in self.PINS_IN_USE:
            GPIO.setup(pin,GPIO.IN)
            self.PINS_IN_USE.append(pin)

        # Read the pin
        resp["message"] = GPIO.input(pin)

        return jsonify(resp)

    def api_lradc_data(self,mode,dat,req_method,req_args):
        resp = copy.deepcopy(self.CHIP_INFO)
        resp["connected"] = True
        resp["mode"] = mode
        # Get Sample Rate
        if mode == "sample_rate" and dat == None and req_method == 'GET':
            resp["message"] = LRADC.get_sample_rate()
        # Set Sample Rate
        elif mode == "sample_rate" and dat != None and req_method in ['GET','PUT','POST']:
            dat = float(dat)
            if dat in [32.25,62.5,125,250]:
                resp["message"] = "Setting LRADC Sample Rate to " + str(dat)
                LRADC.set_sample_rate(dat)
        # Scale Factor
        elif mode == "scale_factor" and req_method == 'GET':
            resp["message"] = LRADC.get_scale_factor()
        # Get Data
        elif (mode == "full" or mode == "raw") and req_method == 'GET':
            dat = int(dat)
            if dat not in [0,1]:
                resp["message"] = "Invalid ADC Channel Specified"
            elif dat == 0:
                if mode == "full":
                    resp["message"] = LRADC.get_chan0()
                elif mode == "raw":
                    resp["message"] = LRADC.get_chan0_raw()
            elif dat == 1:
                if mode == "full":
                    resp["message"] = LRADC.get_chan1()
                elif mode == "raw":
                    resp["message"] = LRADC.get_chan1_raw()
        else:
            resp["message"] = "invalid command"
        return jsonify(resp)

    def api_unexport_all_pins():
        UT.unexport_all()
        resp = copy.deepcopy(self.CHIP_INFO)
        resp["connected"] = True
        resp["message"] = "Unexporting all the Pins"
        return jsonify(resp)

    def api_handle_1v8pin(command,voltage=None):
        resp = copy.deepcopy(self.CHIP_INFO)
        resp["connected"] = True

        # If the command is "voltage" we are requesting the current voltage setting
        if command == "voltage" and not UT.is_chip_pro():
            resp["message"] = "1.8V Pin Current Voltage: " + str(UT.get_1v8_pin_voltage())
        # Disable the 1v8 Pin
        elif command == "disable" and not UT.is_chip_pro():
            resp["message"] = "Disabling the 1.8V Pin"
            UT.disable_1v8_pin()
        elif command == "enable" and not UT.is_chip_pro():
            # Enable the 1v8 Pin
            voltage = float(voltage)
            if voltage not in [1.8, 2.0, 2.6, 3.3]:
                resp["message"] = "invalid voltage specified"
            else:
                resp["message"] = "Enabling the 1.8V Pin to " + str(voltage)
                UT.set_1v8_pin_voltage(voltage)
        else:
            resp["message"] = "invalid command"

        return jsonify(resp)

    def api_pwm(self,chan,command,option,req_method,req_args):
        resp = copy.deepcopy(self.CHIP_INFO)
        resp["connected"] = True

        # Default the channel to PWM0
        # CHIP Pro will support PWM1
        cname = "PWM0"

        chan = int(chan)
        if chan not in [0]: #,1]:
            resp["message"] = "Invalid PWM Channel Specified"
            return jsonify(resp)
        else:
            if chan == 0:
                cname = "PWM0"
            elif chan == 1:
                cname = "PWM1"

        # Figure out our command
        if command == "start" and req_method == 'GET':
            # Load the overlay
            if not UT.is_chip_pro():
                OM.load(cname)
            # Get the arguments
            duty_cycle = req_args.get('duty_cycle', 25.0)
            frequency = req_args.get('frequency', 200.0)
            polarity = req_args.get('polarity', 0)
            # Start the PWM
            PWM.start(cname,duty_cycle,frequency,polarity)
            resp["message"] = "Setting {0} to duty cycle: {1}, frequency: {2}, and polarity {3}".format(cname,duty_cycle,frequency,polarity)
        elif command == "stop" and req_method == 'GET':
            PWM.stop(chame)
            resp["message"] = "Stopping {0}".format(cname)
        elif command == "cleanup" and req_method == 'GET':
            # TODO: Handle per channel cleanup
            PWM.cleanup()
            if not UT.is_chip_pro():
                OM.unload(cname)
            resp["message"] = "Cleaning up and unloading {0}".format(cname)
        elif command == "duty_cycle" and req_method in ['GET','PUT','POST']:
            PWM.set_duty_cycle(cname, float(option))
            resp["message"] = "Changing duty cycle on {0} to {1}".format(cname,option)
        elif command == "frequency" and req_method in ['GET','PUT','POST']:
            PWM.set_frequency(cname, float(option))
            resp["message"] = "Changing frequency on {0} to {1}".format(cname,option)
        return jsonify(resp)
        
    def api_softpwm(self,pin,command,option,req_method,req_args):
        resp = copy.deepcopy(self.CHIP_INFO)
        resp["connected"] = True

        # Figure out our command
        if command == "start" and req_method == 'GET':
            # Get the arguments
            duty_cycle = req_args.get('duty_cycle', 25.0)
            frequency = req_args.get('frequency', 35.0)
            polarity = req_args.get('polarity', 0)
            # Start the SoftPWM
            SPWM.start(pin,duty_cycle,frequency,polarity)
            resp["message"] = "Setting {0} to duty cycle: {1}, frequency: {2}, and polarity {3}".format(pin,duty_cycle,frequency,polarity)
        elif command == "stop" and req_method == 'GET':
            SPWM.stop(pin)
            resp["message"] = "Stopping {0}".format(pin)
        elif command == "cleanup" and req_method == 'GET':
            SPWM.cleanup(pin)
            resp["message"] = "Cleaning up {0}".format(pin)
        elif command == "duty_cycle" and req_method in ['GET','PUT','POST']:
            SPWM.set_duty_cycle(pin, float(option))
            resp["message"] = "Changing duty cycle on {0} to {1}".format(pin,option)
        elif command == "frequency" and req_method in ['GET','PUT','POST']:
            SPWM.set_frequency(pin, float(option))
            resp["message"] = "Changing frequency on {0} to {1}".format(pin,option)
        return jsonify(resp)
    
    def api_servo(self,pin,command,option,req_method,req_args):
        resp = copy.deepcopy(self.CHIP_INFO)
        resp["connected"] = True

        # Figure out the command
        if command == "start" and req_method == 'GET':
            # Get the arguments
            angle = req_args.get('angle', 0.0)
            max_range = req_args.get('max_range', 18.0)
            # Start the servo
            SERVO.start(pin,angle,max_range)
            resp["message"] = "Setting {0} to angle: {1}, max_range: {2},".format(pin,angle,max_range)
        elif command == "stop" and req_method == 'GET':
            SERVO.stop(pin)
            resp["message"] = "Stopping {0}".format(pin)
        elif command == "cleanup" and req_method == 'GET':
            SERVO.cleanup(pin)
            resp["message"] = "Cleaning up {0}".format(pin)
        elif command == "angle" and req_method in ['GET','PUT','POST']:
            SERVO.set_angle(pin, float(option))
            resp["message"] = "Changing angle on {0} to {1}".format(pin,option)
        elif command == "frequency" and req_method in ['GET','PUT','POST']:
            SERVO.set_range(pin, float(option))
            resp["message"] = "Changing max range on {0} to {1}".format(pin,option)
        return jsonify(resp)

# Flask App
app = CHIP_RestAPI(__name__)

# User Functions
def set_id(id):
    app.set_id(id)

def set_name(name):
    app.set_name(name)

def set_hardware(hw):
    app.set_hardware(hw)

def variable(name,value):
    app.set_variable(name, value)

def function(name,funct):
    app.set_function(name,funct)

# ==== The REST App ====
# Actual Rest App
def RestApp(host="0.0.0.0",port=1883,debug=False):
    try:
        app.run(host=host,port=port,debug=debug)
    except KeyboardInterrupt:
        if not UT.is_chip_pro():
            OM.unload("PWM0")
            UT.disable_1v8_pin()
        GPIO.cleanup()
        PWM.cleanup()
        SPWM.cleanup()
        sys.exit(0)

# ==== API Basic Data ====
# Get the basic data
# GET: /
# GET: /id
@app.route('/', methods=['GET'])
@app.route('/id', methods=['GET'])
def index():
    app.CHIP_INFO["connected"] = True
    return jsonify(app.CHIP_INFO)

# Get and Set Variables
# Execute functions
# GET: /<variablename>
# PUT,POST: /<variablename>?value=<value>
# GET: /<functionname>
# GET: /<functionname>?value=<value>
@app.route('/<string:variable>', methods=['GET','PUT','POST','DELETE'])
def get_variables(variable=None):
    return app.api_get_variables(variable,request.method,request.args)

# ==== API DEBUG ====
# GET: /debug
@app.route('/debug', methods=['GET'])
def get_api_debug():
    return app.api_debug()

# ==== CHIP_IO Basics ====
# Get the CHIP_IO Version
# GET: /version
@app.route('/version', methods=['GET'])
def get_chipio_version():
    return app.api_chipio_version()

# ==== GPIO ====
# Digital Read and Cleanup
# GET: /digital/cleanup
# GET: /digital/cleanup/<pinname>
# GET: /digital/<pinname>/r
# GET: /digital/<pinname>
@app.route('/digital/<string:command>', defaults={'pin' : None}, methods=['GET'])
@app.route('/digital/<string:command>/<string:pin>', methods=['GET'])
def digital_read_cleanup(command,pin=None):
    # We want to cleanup
    if command == "cleanup":
        # If we have no mode specified (mode == pin)
        if pin == None:
            return app.api_digital_cleanup()
        # Do the pin cleanup, if the pin value isn't "r" for a read
        elif pin != "r":
            return app.api_digital_pin_cleanup(pin)
    else:
        # For the read, the pin name is the command
        return app.api_digital_read(command)

# Digital Write
# GET,PUT,POST: /digital/<pinname>/[0,1]
@app.route('/digital/<string:pin>/<int:value>', methods=['GET','PUT','POST'])
def digital_write_command(pin,value):
    return app.api_digital_write(pin,value)

# ==== PWM ====
# GET: /pwm/[0,1]/start?duty_cycle=<dutycycle>&frequency=<frequency>&polarity=[0,1]
# GET: /pwm/[0,1]/stop
# GET: /pwm/[0,1]/cleanup
# GET, PUT, POST: /pwm/[0,1]/duty_cycle/<dutycycle>
# GET, PUT, POST: /pwm/[0,1]/frequency/<frequency>
@app.route('/pwm/<int:chan>/<string:command>', methods=['GET','PUT','POST'])
@app.route('/pwm/<int:chan>/<string:command>/<string:option>', methods=['GET','PUT','POST'])
def pwm_all_commands(chan,command,option=None):
    return app.api_pwm(chan,command,option,request.method,request.args)

# ==== SOFTPWM ====
# GET: /softpwm/<pinname>/start?duty_cycle=<dutycycle>&frequency=<frequency>&polarity=[0,1]
# GET: /softpwm/<pinname>/stop
# GET: /softpwm/<pinname>/cleanup
# GET, PUT, POST: /softpwm/<pinname>/duty_cycle/<dutycycle>
# GET, PUT, POST: /softpwm/<pinname>/frequency/<frequency>
@app.route('/softpwm/<string:pin>/<string:command>', methods=['GET','PUT','POST'])
@app.route('/softpwm/<string:pin>/<string:command>/<string:option>', methods=['GET','PUT','POST'])
def softpwm_all_commands(pin,command,option=None):
    return app.api_softpwm(pin,command,option,request.method,request.args)

# ==== SERVO ====
# Methods
# GET: /servo/<pinname>/start?angle=<angle>&max_range=<range>
# GET: /servo/<pinname>/stop
# GET: /servo/<pinname>/cleanup
# GET, PUT, POST: /servo/<pinname>/angle/<angle>
# GET, PUT, POST: /servo/<pinname>/max_range/<range>
@app.route('/servo/<string:pin>/<string:command>', methods=['GET','PUT','POST'])
@app.route('/servo/<string:pin>/<string:command>/<string:option>', methods=['GET','PUT','POST'])
def servo_all_commands(pin,command,option=None):
    return api.api_servo(pin,command,option,request.method,request.args)

# ==== LRADC ====
# Methods
# GET: /analog/sample_rate
# GET: /analog/scale_factor
# GET: /analog/full/[0,1]
# GET: /analog/raw[0,1]
# GET,PUT,POST: /analog/sample_rate/[32.25,62.5,125,250]
@app.route('/analog/<mode>', methods=['GET'])
@app.route('/analog/<string:mode>/<string:dat>', methods=['GET','PUT','POST'])
def get_lradc_data(mode,dat=None):
    return app.api_lradc_data(mode,dat,request.method,request.args)

# ==== Utilities ====
# Methods
# GET: /unexport_all
@app.route('/unexport_all', methods=['GET'])
def unexport_all_pins():
    return app.api_unexport_all_pins()

# GET: /1v8_pin/voltage
# GET: /1v8_pin/disable
# GET,PUT,POST: 1v8_pin/enable/[1.8,2.0,2.6,3.3]
@app.route('/1v8_pin/<string:command>', methods=['GET'])
@app.route('/1v8_pin/<string:command>/<float:voltage>', methods=['GET','PUT','POST'])
def handler_1v8pin(command,voltage=None):
    return app.api_handle_1v8_pin()

# DEBUG Testing
if __name__ == '__main__':
    RestApp(debug=True)

